### PR TITLE
 Add Weighted Loss Functions to PyTorch : WMSE, WMAE, and Weighted Huber Loss

### LINKED ISSUE #132465: Add Weighted Loss Functions to PyTorch : WMSE, WMAE, and Weighted Huber Loss
### 🚀 The feature, motivation and pitch

The addition of weighted loss functions to the PyTorch library, specifically Weighted Mean Squared Error (WMSE), Weighted Mean Absolute Error (WMAE), and Weighted Huber Loss, significantly enhances the library’s capability to handle imbalanced datasets and prioritize specific samples during training. In many real-world scenarios, datasets are not perfectly balanced; some samples are inherently more important or represent edge cases that require greater emphasis during the training process. Traditional loss functions treat all samples equally, which can lead to suboptimal models, especially in the presence of noise or outliers.

By incorporating weighted loss functions, users gain precise control over the influence of each sample. This allows for improved model performance on imbalanced datasets by emphasizing more critical samples and downplaying less significant ones. For instance, in medical image analysis, certain images might be more crucial for diagnosing rare conditions, or in financial forecasting, particular time periods might hold more significance due to market volatility. These functions empower users to tailor their loss calculations to the specific needs of their data, resulting in more robust and accurate models.

### Alternatives

Several alternative approaches and features were considered before deciding to implement weighted loss functions:

1. **Data Augmentation and Sampling Techniques:**
   - One alternative is to use data augmentation or resampling techniques to balance the dataset. While these methods can help, they often require additional preprocessing steps and can lead to overfitting or increased computational overhead.

2. **Custom Loss Functions:**
   - Users could implement their custom loss functions by manually applying weights. However, this approach can be error-prone and less efficient compared to having optimized, built-in functions within the library.

3. **Class Weighting in Existing Loss Functions:**
   - For classification tasks, many loss functions already support class weighting. However, this does not extend to regression tasks or more nuanced scenarios where per-sample weighting is necessary.

4. **Other Libraries and Extensions:**
   - While other machine learning frameworks or third-party extensions might offer similar functionality, integrating these capabilities directly into PyTorch ensures better compatibility, ease of use, and performance optimization.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tolleybot**
#### Summary
This pull request introduces new weighted loss functions to the PyTorch library: `weighted_huber_loss`, `wmse_loss`, and `wmae_loss`. These functions allow for precise control over the influence of each sample during training, important for imbalanced data or when certain samples are more significant than others.

#### Changes
- **`weighted_huber_loss`**: Huber loss modified to incorporate weights, providing a balance between L1 and L2 loss based on the `delta` parameter.
- **`wmse_loss`** (Weighted Mean Squared Error): Applies weights to the standard MSE loss, useful for emphasizing certain samples in regression tasks.
- **`wmae_loss`** (Weighted Mean Absolute Error): Adjusts MAE loss calculation by including weights, ideal for datasets with outliers.

#### Code Details
- **Input Validation**: Ensures `input`, `target`, and `weights` tensors match in size to prevent broadcasting errors.
- **Reduction Options**: Supports `none`, `mean`, and `sum` reductions to suit various computational needs.
- **Backward Compatibility**: Maintains support for deprecated arguments `size_average` and `reduce`, while encouraging use of the `reduction` argument.

#### Usage Example
```python
import torch
input = torch.tensor([0.5, 2.5, 2.0], dtype=torch.float32)
target = torch.tensor([0.0, 2.0, 1.5], dtype=torch.float32)
weights = torch.tensor([1.0, 0.5, 1.5], dtype=torch.float32)

loss = weighted_huber_loss(input, target, weights, delta=1.0)
print(loss)
```
---

Feedback on these implementations is welcome; please let me know if further modifications are required.

Resolves #132465

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### CODE REVIEW

**Comment by jbschlosser** [comment]
Hi @tolleybot, is there an associated issue for this PR?

**Review by mikaylagawarecki — COMMENTED** [review]
See comment on issue https://github.com/pytorch/pytorch/issues/132465#issuecomment-2266156358

**Comment by mikaylagawarecki** [comment]
seems like tests are failing

File: torch/nn/functional.py [hunk 1]
```diff
@@ -3762,36 +3802,145 @@ def mse_loss(
...
+                f"Invalid reduction mode: {reduction}. Expected one of 'none', 'mean', 'sum'."
+            )
+    else:
+        expanded_input, expanded_target = torch.broadcast_tensors(input, target)
+        return torch._C._nn.mse_loss(
+            expanded_input, expanded_target, _Reduction.get_enum(reduction)
+        )
+
+
+def mae_loss(
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:3874]: Why are we adding mae_loss, isn't this synonymous to l1_loss? (from my previous comment, for wmae_loss we can add the weight argument to l1_loss), let me know if I missed something however :)

Comment (tolleybot) [reply torch/nn/functional.py:3874]: No problem, let me take a look.  If its redudent then I'll remove it

Comment (tolleybot) [reply torch/nn/functional.py]: I have removed mae_loss and just added an optional weight param to l1_loss.

**Comment by adamreeve** [comment]
Hi @mikaylagawarecki, can you please take another look at this when you have time, is there anything else needed before this can be merged?

File: torch/nn/functional.py [hunk 2]
```diff
@@ -3683,26 +3683,41 @@ def huber_loss(
     target: Tensor,
     reduction: str = "mean",
     delta: float = 1.0,
+    weights: Optional[Tensor] = None,
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:3686]: Please name these `weight` rather than `weights` for consistency with other loss functions with this argument in this file, similarly in the other functions

File: torch/nn/functional.py [hunk 3]
```diff
@@ -3712,9 +3727,34 @@ def huber_loss(
...
+
+        # Apply weights to the unweighted loss
+        weighted_loss = unweighted_loss * weights
+
+        if reduction == "none":
+            return weighted_loss
+        elif reduction == "sum":
+            return torch.sum(weighted_loss)
+        elif reduction == "mean":
+            return torch.sum(weighted_loss) / torch.sum(weights)
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:3753]: hm, why is this not `weighted_loss.mean()`? which seems to be the precedent for bce_loss with weights as well as multimarginloss with weights in this file, similarly below https://github.com/pytorch/pytorch/blob/46935c8241db2df5ea2d5da6025d53b343f3e094/torch/nn/functional.py#L3958-L3959

File: test/test_nn.py [hunk 4]
```diff
@@ -12993,7 +13024,7 @@ def __init__(self) -> None:
         self.assertEqual(list(state_dict._metadata.keys()), list(ddp_state_dict._metadata.keys()))
 
 
-instantiate_device_type_tests(TestNNDeviceType, globals(), allow_mps=True)
+instantiate_device_type_tests(TestNNDeviceType, globals())
```
Comment (mikaylagawarecki) [inline test/test_nn.py:13027]: why did you remove `allow_mps` here? Please do not do this

It seems like your linter also proposed a bunch of changes, could you undo these or split them into a separate PR please?

File: torch/nn/functional.py [hunk 5]
```diff
@@ -3683,26 +3683,41 @@ def huber_loss(
     target: Tensor,
     reduction: str = "mean",
     delta: float = 1.0,
+    weights: Optional[Tensor] = None,
 ) -> Tensor:
-    r"""Compute the Huber loss.
+    r"""huber_loss(input, target, reduction='mean', delta=1.0, weights=None) -> Tensor
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:3688]:
```suggestion
    r"""huber_loss(input, target, reduction='mean', delta=1.0, weight=None) -> Tensor
```

File: torch/nn/functional.py (hunk 2 above)
Comment (mikaylagawarecki) [inline torch/nn/functional.py:3686]:
```suggestion
    weight: Optional[Tensor] = None,
```

**Review by mikaylagawarecki — COMMENTED** [review]
Apologies for the delay, this PR fell out of my queue

File: torch/nn/functional.py [hunk 6]
```diff
@@ -3683,26 +3683,41 @@ def huber_loss(
...
+        weights (Tensor, optional): Weights for each sample. Default: None.
+
+    Returns:
+        Tensor: Huber loss (optionally weighted).
     """
     if has_torch_function_variadic(input, target):
         return handle_torch_function(
-            huber_loss,
-            (input, target),
+            huber_loss(input, target, weights),
```
Comment (mikaylagawarecki) [inline torch/nn/functional.py:3713]:
```suggestion
            huber_loss,
            (input, target, weights),
```

File: test/test_nn.py (hunk 4 above)
Comment (tolleybot) [reply test/test_nn.py:13027]: will do

File: torch/nn/functional.py (hunk 5 above)
Comment (tolleybot) [reply torch/nn/functional.py]: Done

File: torch/nn/functional.py (hunk 2 above)
Comment (tolleybot) [reply torch/nn/functional.py]: Done

File: torch/nn/functional.py (hunk 3 above)
Comment (tolleybot) [reply torch/nn/functional.py]: Makes sense!

**Comment by mikaylagawarecki** [comment]
Two code pointers that might help https://github.com/pytorch/pytorch/blob/main/torch//overrides.py#L971 and https://github.com/pytorch/pytorch/blob/main/torch/_refs/nn/functional/__init__.py?fbclid=IwZXh0bgNhZW0CMTEAAR1OLkVMxcuGxt87Q_3K1uuoFkz3ZwOL3EUB7CWOwnxVCMHaPipRO9bqbKw_aem_i7dF-atGh9GtEuU1a4ycgA#L862

**Comment by mikaylagawarecki** [comment]
Hey referring to my [previous comment](https://github.com/pytorch/pytorch/pull/132049#issuecomment-2354253606), I'm pretty sure you need to update the [refs](url) ([second link](https://github.com/pytorch/pytorch/blob/main/torch/_refs/nn/functional/__init__.py?fbclid=IwZXh0bgNhZW0CMTEAAR1OLkVMxcuGxt87Q_3K1uuoFkz3ZwOL3EUB7CWOwnxVCMHaPipRO9bqbKw_aem_i7dF-atGh9GtEuU1a4ycgA#L862)) as well

**Review by mikaylagawarecki — APPROVED** [review]
Thanks
